### PR TITLE
refactor(doc): improvements to hacking.rst

### PR DESCRIPTION
**Pull request by Alizter**
I have:

- Made sure line wrapping is consistent.
- Added a table of contents.
- Added information about Nix environments for devs.

### CODE REVIEW

File: doc/hacking.rst
Comment (Alizter) [reply doc/hacking.rst:135]: Should we add `nix` to `.gitignore` then?

Comment (rgrinberg) [reply doc/hacking.rst:135]: no, we'll want to keep our nix library code there.

Comment (Alizter) [reply doc/hacking.rst:135]: What about `nix/profiles`?

Comment (rgrinberg) [reply doc/hacking.rst:135]: That one should be ignored because profiles shouldn't be committed

Comment (Alizter) [reply doc/hacking.rst]: Ah it already is ignored, my bad.

File: doc/hacking.rst
Comment (christinerose) [inline doc/hacking.rst:23]:
```suggestion
invocation of ``ocamlopt`` or ``ocamlc`` via the ``bootstrap.ml`` OCaml script.
```

File: doc/hacking.rst
Comment (christinerose) [inline doc/hacking.rst:122]:
```suggestion
Setting Up Your Development Environment Using Nix
```

**Review by christinerose — APPROVED** [review]
Some minor suggestions in formatting/grammar, but I'm approving it so you don't have to wait on me to approve before merging. 😄

**Review by rgrinberg — APPROVED** [review]

**Pull request merged**
